### ISSUE TITLE
A question about paper

### ISSUE THREAD

**Issue by liuxinbin12138** [spcl/MRAG#1]
I'm curious, the paper uses a decoder for the embedding model, which is unusual since encoders or bi-encoders are more commonly used. Could you help me figure out why they went with that choice?

**Comment by rgersten** [comment]
We did not choose the Salesforce/SFR-Embedding-Mistral model because of its architectural features, but because it was at the top of MTEB leaderboard (https://huggingface.co/spaces/mteb/leaderboard) at the time of the decision.

**Comment by liuxinbin12138** [comment]
> We did not choose the Salesforce/SFR-Embedding-Mistral model because of its architectural features, but because it was at the top of MTEB leaderboard (https://huggingface.co/spaces/mteb/leaderboard) at the time of the decision.

Thx for your help!

**Comment by liuxinbin12138** [comment]
If we use another encoder (e.g., BAAI/bge-base-en-v1.5), the following code does not work, how can I fix it? could you please give some suggestions?    

def generate_embeddings(self, text: str) -> FullEmbeddings:
        """
        Generate embeddings (standard embedding, attention head embeddings) for the input text.

        :param text: Input text.
        :type text: str
        :return: Embeddings.
        :rtype: FullEmbeddings
        """
        all_layer_embeddings: dict[int, LayerEmbeddings] = {}

        def split(_embedding: Embedding, _interval: int) -> list[Embedding]:
            _sub_embeddings: list[Embedding] = []
            for i in range(0, len(_embedding), _interval):
                _sub_embeddings.append(_embedding[i:(i + _interval)])
            return _sub_embeddings

        with torch.no_grad():
            batch_dict = self._tokenizer([text], padding=True, return_tensors="pt")
            iids = batch_dict["input_ids"].to(self._model.device)
            hidden_states = self._model.embed_tokens(iids)

            for layer_idx, layer in enumerate(self._model.layers):
                # run inference for this layer
                hidden_states = layer(hidden_states)[0]

                if layer_idx in self.target_layers:
                    # retrieve cached input from custom module
                    attn_heads = layer.self_attn.o_proj.last_input[0][-1]
                    layer_embeddings = LayerEmbeddings(
                        attention_heads=split(attn_heads.tolist(), 128),
                    )
                    all_layer_embeddings[layer_idx] = layer_embeddings

            # apply last layer, followed by the RMS norm
            standard_embedding = self._model.norm(hidden_states)[0][-1].tolist()
            return FullEmbeddings(standard_embedding, all_layer_embeddings)

**Comment by rgersten** [comment]
It is not enough to just change the model name, the EmbeddingModel class has also be updated to the specific architecture of the model in question. If you manage to implement support for "BAAI/bge-base-en-v1.5", please consider creating a PullRequest, so that we can integrate your code into our repository.

We are currently working on adding support for additional embedding models, but not the model that you want to use, to our evaluation pipeline and will release a new version in the near future.

By the way, in response to your earlier question, in one of our other recent publications (https://arxiv.org/abs/2406.02524), we showed that decoder based encodings are pretty powerful and can be used in the embedding context as well (this is in line with the MTEB leaderboard, where most leading embedding models are based on the decoder-only architecture).
